qutip/solver/mcsolve.py

```diff
@@ -0,0 +1,311 @@
+        (atol, rtol) for each e_ops.
+
+    timeout : float [optional]
+        Maximum time for the evolution in second. When reached, no more
+        trajectories will be computed. Overwrite the option of the same name.
+
+    Returns
+    -------
+    results : :class:`qutip.solver.Result`
+        Object storing all results from the simulation. Which results is saved
+        depends on the presence of ``e_ops`` and the options used. ``collapse``
+        and ``photocurrent`` is available to Monte Carlo simulation results.
+    """
+    H = QobjEvo(H, args=args, tlist=tlist)
+    if not isinstance(c_ops, (list, tuple)):
+        c_ops = [c_ops]
+    c_ops = [QobjEvo(c_op, args=args, tlist=tlist) for c_op in c_ops]
+
+    if len(c_ops) == 0:
+        return mesolve(H, state, tlist, e_ops=e_ops, args=args,
+                       options=options)
+
+    if isinstance(ntraj, list):
+        raise TypeError("No longer supported, use `result.expect_traj_avg`"
+                        "with the options `keep_runs_results=True`.")
+
+    mc = McSolver(H, c_ops, options=options)
+    result = mc.run(state, tlist=tlist, ntraj=ntraj, e_ops=e_ops,
+                    seed=seeds, target_tol=target_tol, timeout=timeout)
+    return result
+
+
+class _McTrajectorySolver(_TrajectorySolver):
+    """
+    Solver for one mcsolve trajectory. Created by a :class:`McSolver`.
+    """
+    name = "mcsolve"
+    _avail_integrators = {}
+
+    def __init__(self, rhs, c_ops, n_ops, *, options=None, _prepare=False):
+        self._c_ops = c_ops
+        self._n_ops = n_ops
+        super().__init__(rhs, options=options, _prepare=_prepare)
+
+    def _run(self, seed, state, tlist, e_ops):
+        self.collapses = []
+        self.generator = self.get_generator(seed)
+        self.target_norm = self.generator.random()
+        result = super()._run(self.generator, state, tlist, e_ops)
+        result.collapse = list(self.collapses)
+        result.seed = seed
+        return result
+
+    def start(self, state, t0, seed=None, *, safe_ODE=None):
+        self.collapses = []
+        self.generator = self.get_generator(seed)
+        self.target_norm = self.generator.random()
+        super().start(state, t0, seed=self.generator, safe_ODE=safe_ODE)
+
+    def prob_func(self, state):
+        if self.rhs.issuper:
+            return _data.norm.trace(unstack_columns(state))
+        return _data.norm.l2(state)**2
+
+    def norm_func(self, state):
+        if self.rhs.issuper:
+            return _data.norm.trace(unstack_columns(state))
+        return _data.norm.l2(state)
+
+    def _step(self, t, copy=True):
+        t_old, y_old = self._integrator.get_state(copy=False)
+        norm_old = self.prob_func(y_old)
+        while t_old < t:
+            t_step, state = self._integrator.mcstep(t, copy=False)
+            norm = self.prob_func(state)
+            if norm <= self.target_norm:
+                self._do_collapse(norm_old, norm, t_old)
```

Comment:
May I suggest splitting this in two? one that finds the collapse (`_find_collapse_time`) time and another that applies the collapse . These operations by themselves are relatively complex so I think it will help in the readability/maintanability. Also, I suggest using as argument names for `_find_collapse_time(norm_t_old, t_old, norm_t, t)` so that it is clear that `norm_t_old` refers to the norm at `t_old` whereas `norm_t` refers to the norm at time t. At this moment t is not a parameter and although it can be obtain from integrator I would say it is better to pass it in order to be explicit on what the function does (I believe this will help writing a clear docstring).